src/hal/src/external_memory/fd.rs

```diff
@@ -0,0 +1,144 @@
+use super::{ExternalMemory, ExternalMemoryType, ExternalMemoryTypeFlags};
+
+/// Unix file descriptor
+#[derive(Debug)]
+pub struct Fd(i32);
+impl From<i32> for Fd {
+    fn from(fd: i32) -> Self {
+        Self(fd)
+    }
+}
+impl std::os::unix::io::AsRawFd for Fd {
+    fn as_raw_fd(&self) -> std::os::unix::io::RawFd {
+        self.0
+    }
+}
+impl std::ops::Deref for Fd {
+    type Target = i32;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+#[cfg_attr(feature = "unstable", doc(cfg(unix)))]
+#[derive(Debug)]
+#[allow(non_camel_case_types)]
+/// External memory that rely on unix file descriptors
+pub enum ExternalMemoryFd {
```

Comment:
Does it make sense to separate the different external memory types? Can there instead be one `enum` with both Unix FD's, DXGI handles, and Apple's IO surfaces?

Reply:
Since, with the new code, all platform specific functions have been merged, most of those "subtypes" are no more needed and they have been removed. There are a couple of types that are used in the vulkan backend to make possible some "match arms" to share code path, so reducing the code size. I will try to remove them too if possible.